### PR TITLE
kubelet: refactor kubecontainer.Pod, remove PodStatus from it.

### LINKED ISSUE #12619: Some thoughts about the pod/container abstraction in kubelet
This is purely food for thought, as others may not agree and this probably wouldn't be prioritized anytime soon :-)

Kubelet defines [`Pod`](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/container/runtime.go#L117) in the container runtime, which has and a list of [containers](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/container/runtime.go#L170) and an [`api.PodStatus`](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/container/runtime.go#L131) object which includes all the container statuses. This seems counter-intuitive to me because `Pod` is supposed to be a representation of the running pod in the system, i.e., it is essentially the pod status/states (so we have status nested in the status?) It is also not necessary for the container runtime to talk `api.PodStatus` because kubelet has to [fill extra information later on to complete it](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L2326-L2370). 

The abstraction is a bit hard to digest alone, and the types seem to center around the implementation details (retrofitting?). Moreover, we pass Pods often without any valid pod/container status. You have no way to tell after-the-fact that whether a container in the pod is running or dead. I'd prefer a flatter struct reflecting the status of the containers/pods, and let kubelet interprets and translates the status into `api.PodStatus`.

A simple example would be:

```go
type Container struct {
	ID  types.UID
	Name string
	Image string
	Hash uint64
    State ContainerState
    Ready bool
        ...
}
type Pod struct {
	ID types.UID
	Name      string
	Namespace string
	Containers []*Container
    PodIP string
}
```

/cc @yifan-gu

### PR DESCRIPTION
**Pull request by yifan-gu**
Related to #12619
Remove the PodStatus in kubecontainer.Pod.

### CODE REVIEW

File: pkg/kubelet/rkt/rkt.go [hunk 1]
```diff
@@ -684,18 +673,78 @@ func (r *runtime) KillPod(pod kubecontainer.Pod) error {
...
+		return nil, err
+	}
+	status := podInfo.toPodStatus(pod)
+	return &status, nil
+}
+
+// GetPodStatus returns the status of the given pod.
 func (r *runtime) GetPodStatus(pod *api.Pod) (*api.PodStatus, error) {
-	pods, err := r.GetPods(true)
+	unitNames, err := listUnitFiles()
```
Comment (vishh) [inline pkg/kubelet/rkt/rkt.go]: Listing all unit files on every GetPodStatus() call seems very expensive considering that we poll for Pod Status periodically as of now. Are there any alternatives to this approach? Inotify or cache unit files relevant to a pod, etc?

**Comment by vishh** [comment]
ok to test

**Comment by vishh** [comment]
@yifan-gu: LGTM except for the one comment.

File: pkg/kubelet/rkt/rkt.go (hunk 1 above)
Comment (yifan-gu) [inline pkg/kubelet/rkt/rkt.go]: @vishh Thanks for catching, yes, this involes a bunch of calls to `stat`. I will add a TODO here to optimize later. (e.g. list unit files at the first time when kubelet starts, and record every creation of the unit file names later, we just need the names here.)

Comment (vishh) [inline pkg/kubelet/rkt/rkt.go]: Ok. Thanks!

**Comment by yifan-gu** [comment]
@yujuhong Any feedback on the this? Note I am not adding `ContainerStatus` in `kubecontainer.Container` here, nor the `PodIP` in `kubecontainer.Pod` as AFAICS they are included in the `PodStatus`. Not sure if you have any specific case to add `ContainerStatus` in `kubecontainer.Container`?

File: pkg/kubelet/rkt/rkt.go (hunk 1 above)
Comment (yifan-gu) [inline pkg/kubelet/rkt/rkt.go]: You are welcome :)

**Comment by yujuhong** [comment]
@yifan-gu, my email setting for github is messed up, so I didn't get the notification. I will look at this asap. Thanks!

File: pkg/kubelet/rkt/rkt.go (hunk 1 above)
Comment (yujuhong) [inline pkg/kubelet/rkt/rkt.go]: Docker is in the similar boat as it has to do to `docker ps -a` to list all containers in order to achieve this. Hopefully adding the runtime pod cache would solve this problem :)

**Comment by yujuhong** [comment]
> @yujuhong Any feedback on the this? Note I am not adding ContainerStatus in kubecontainer.Container here, nor the PodIP in kubecontainer.Pod as AFAICS they are included in the PodStatus. But maybe you have some specific reason to add ContainerStatus in kubecontainer.Container?


@yifan-gu, I don't necessary want the ContainerStatus here, but I'd like `Pod` to include all the information necessary for kubelet to generate api.PodStatus by itself.
That way, if we implement a runtime pod cache, whenever there is a change in the pod/container state, we will inspect the container and update the cache entry. Kubelet can always look at the cache and figure out the api.PodStatus itself, without having to ask individual runtimes. 

In this case, I think you are right *not* to include ContainerStatus directly, but maybe we can think about adding fields like `FinishedAt`, `ExitCode`, etc. Another great reason to do this is that we can GC the container after we update the cache so that we don't have to keep so many containers around.

**Comment by yujuhong** [comment]
Removing PodStatus from Pod is always good. This PR LGTM as it is. We can talk about whether we want to add extra fields later :)
Thanks!

**Comment by yifan-gu** [comment]
@yujuhong Thanks for the heading up. I am thinking that can we let pleg manage and cache the []podstatus? Maybe we can add things that are missing (podID, podName, podNamespace for podstatus, and container hash for container status) ?

Then we can possibly remove the kubecontainer.Pod and related stuff. (The reason they exist mostly is because previously we wanted to implement the runtime-agnostic interfaces quickly, without bothering changing the api)

**Comment by yifan-gu** [comment]
Anyway, would like to get this merged as my following rkt patches will depend on this :)

**Comment by yujuhong** [comment]
> @yujuhong Thanks for the heading up. I am thinking that can we let pleg manage and cache the []podstatus? Maybe we can add things that are missing (podID, podName, podNamespace for PodStatus, and container hash for ContainerStatus) ?
> 
> Then we can possibly remove the kubecontainer.Pod and related stuff. (The reason they exist mostly is because previously we wanted to implement the runtime-agnostic interfaces quickly, without bothering changing the api objects)


I think there are still merits of keeping kubeocntainer.Pod and types alike because the pod cache would be cleaner and it'd be easier to interface with other internal components. api.PodStatus has too many things that doesn't make sense for the underlying runtime. We'll also not be subject to how api.PodStatus is defined in this case :)

**Pull request merged**